src/node_mutex.h

```diff
@@ -0,0 +1,187 @@
+  };
+
+  class ScopedUnlock {
+   public:
+    inline explicit ScopedUnlock(const ScopedLock& scoped_lock);
+    inline ~ScopedUnlock();
+
+   private:
+    friend class ScopedLock;
+    const MutexBase& mutex_;
+    DISALLOW_COPY_AND_ASSIGN(ScopedUnlock);
+  };
+
+ private:
+  template <typename> friend class ConditionVariableBase;
+  mutable typename Traits::MutexT mutex_;
+  DISALLOW_COPY_AND_ASSIGN(MutexBase);
+};
+
+template <typename Traits>
+class ConditionVariableBase {
+ public:
+  using ScopedLock = typename MutexBase<Traits>::ScopedLock;
+
+  inline ConditionVariableBase();
+  inline ~ConditionVariableBase();
+  inline void Broadcast(const ScopedLock&);
+  inline void Signal(const ScopedLock&);
+  inline void Wait(const ScopedLock& scoped_lock);
+
+ private:
+  typename Traits::CondT cond_;
+  DISALLOW_COPY_AND_ASSIGN(ConditionVariableBase);
+};
+
+struct LibuvMutexTraits {
+  using CondT = uv_cond_t;
+  using MutexT = uv_mutex_t;
+
+  static inline int cond_init(CondT* cond) {
+    return uv_cond_init(cond);
+  }
+
+  static inline int mutex_init(MutexT* mutex) {
+    return uv_mutex_init(mutex);
+  }
+
+  static inline void cond_broadcast(CondT* cond) {
+    uv_cond_broadcast(cond);
+  }
+
+  static inline void cond_destroy(CondT* cond) {
+    uv_cond_destroy(cond);
+  }
+
+  static inline void cond_signal(CondT* cond) {
+    uv_cond_signal(cond);
+  }
+
+  static inline void cond_wait(CondT* cond, MutexT* mutex) {
+    uv_cond_wait(cond, mutex);
+  }
+
+  static inline void mutex_destroy(MutexT* mutex) {
+    uv_mutex_destroy(mutex);
+  }
+
+  static inline void mutex_lock(MutexT* mutex) {
+    uv_mutex_lock(mutex);
+  }
+
+  static inline void mutex_unlock(MutexT* mutex) {
+    uv_mutex_unlock(mutex);
+  }
+};
+
+template <typename Traits>
+ConditionVariableBase<Traits>::ConditionVariableBase() {
+  CHECK_EQ(0, Traits::cond_init(&cond_));
+}
+
+template <typename Traits>
+ConditionVariableBase<Traits>::~ConditionVariableBase() {
+  Traits::cond_destroy(&cond_);
+}
+
+template <typename Traits>
+void ConditionVariableBase<Traits>::Broadcast(const ScopedLock&) {
+  Traits::cond_broadcast(&cond_);
+}
+
+template <typename Traits>
+void ConditionVariableBase<Traits>::Signal(const ScopedLock&) {
+  Traits::cond_signal(&cond_);
+}
+
+template <typename Traits>
+void ConditionVariableBase<Traits>::Wait(const ScopedLock& scoped_lock) {
+  Traits::cond_wait(&cond_, &scoped_lock.mutex_.mutex_);
+}
+
+template <typename Traits>
+MutexBase<Traits>::MutexBase() {
+  CHECK_EQ(0, Traits::mutex_init(&mutex_));
+}
+
+template <typename Traits>
+MutexBase<Traits>::~MutexBase() {
+  Traits::mutex_destroy(&mutex_);
+}
+
+template <typename Traits>
+void MutexBase<Traits>::Lock() {
+  Traits::mutex_lock(&mutex_);
+}
+
+template <typename Traits>
+void MutexBase<Traits>::Unlock() {
+  Traits::mutex_unlock(&mutex_);
+}
+
+template <typename Traits>
+MutexBase<Traits>::ScopedLock::ScopedLock(const MutexBase& mutex)
```

Comment:
I can add code that prohibits directly heap-allocating it (see below) but that can still be subverted by using `::new Mutex::ScopedLock(...)` or by embedding the ScopedLock in a heap-allocated struct.

If you still think it's a good idea, LMK and I'll follow up with a pull request.
```diff
diff --git a/src/node_mutex.h b/src/node_mutex.h
index 9e1d316..e14735a 100644
--- a/src/node_mutex.h
+++ b/src/node_mutex.h
@@ -35,6 +35,7 @@ class MutexBase {
     friend class ScopedUnlock;
     const MutexBase& mutex_;
     DISALLOW_COPY_AND_ASSIGN(ScopedLock);
+    DISALLOW_NEW_AND_DELETE();
   };
 
   class ScopedUnlock {
@@ -46,6 +47,7 @@ class MutexBase {
     friend class ScopedLock;
     const MutexBase& mutex_;
     DISALLOW_COPY_AND_ASSIGN(ScopedUnlock);
+    DISALLOW_NEW_AND_DELETE();
   };
 
  private:
diff --git a/src/util.h b/src/util.h
index 9ad0f6c..a3e211a 100644
--- a/src/util.h
+++ b/src/util.h
@@ -33,6 +33,13 @@ template <typename T> using remove_reference = std::remove_reference<T>;
   TypeName(const TypeName&) = delete;                                         \
   TypeName(TypeName&&) = delete
 
+#define DISALLOW_NEW_AND_DELETE()                                             \
+  void* operator new(size_t) = delete;                                        \
+  void* operator new[](size_t) = delete;                                      \
+  void operator delete(void*, size_t) = delete;                               \
+  void operator delete[](void*, size_t) = delete
+
 // Windows 8+ does not like abort() in Release mode
 #ifdef _WIN32
 #define ABORT() raise(SIGABRT)
```